company/okrs/2020_q1.md

```diff
@@ -88,28 +88,18 @@ Build out vision for product and run customer validation discussions
    1. Web KR: At least four customers are using campaigns in production
    1. Web KR: All campaign types have been used by customers in production
 
-1. **Core Services: Build features that accelerate customers through pipeline**
-   1. Core Services KR: Shorten the time from "intent to try Sourcegraph" to "whole org is using Sourcegraph" (onboarding experience)
-   1. Core Services KR: Ensure new features (Automation, ACLs, etc) are accessible and easily tryable by customers once they go GA
-
-1. **Core Services: Search has feature parity with OpenGrok**
-   1. Core Services KR: Implement AND query support
-   1. Core Services KR: Implement OR query support
-   1. Core Services KR: Implement NOT query support
-
-1. **Core Services: Existing customers are cared for**
-   1. Core Services KR: Each iteration work plan includes prioritised customer issues.
-   1. Core Services KR: Each team mate works on relevant customer issues every iteration.
+1. **Core Services: Search gains new capabilities which are adopted by customers**
+   1. Core Services KR: Support AND/OR/NOT operators on file contents.
+   1. Core Services KR: Write at least 2 blog posts about new search features.
+   1. Core Services KR: Structural search can be ran at scale (~30k repositories).
```

Comment:
Everything for search looks great, but I still don't know about the structural search at scale thing. What does it mean, what runtimes are we looking for? I can say that already structural search will be 5x to 10x slower than regex, and this is reasonable because it does something more specialized. Structural search (or any 'analysis') will "run" at 30K repos but could take 'forever'. So without a time bound or idea around that I don't know what this KR means. I suggest we remove it.

Reply:
What if we changed this KR to be something like "add metrics for search latency per search type" (my assumption is that we don't track this from customer instances yet).

Reply:
Alternatively (although we should probably add metrics regardless), I think the point of this KR is to assert that structural search can actually be used at large customers. It wouldn't be useless, for example, if structural searches on 30k repos timed out after 60s.

I think it should be possible to formulate a KR like this:

```suggestion
   1. Core Services KR: Structural search over ~30k repositories returns results in less than X seconds.
```

Where X can be large enough to make you @rvantonder comfortable that this is possible to achieve, but low enough to be reasonable for a human to wait for with the context that they have never had this capability before (e.g. X = 30s might be ok).

A rough estimate would be fine here (e.g. if we expect this to be 10x slower than regex, then how fast are regex queries on 30k repos?).

Reply:
> A rough estimate would be fine here (e.g. if we expect this to be 10x slower than regex, then how fast are regex queries on 30k repos?)

That's a good perspective. I think the trouble with doing this by number of repos is that it conflates speed of structural/regex search and the performance of the infrastructure around dealing with 30K repos, without saying how big those repos are. We can thumbsuck values for how many lines of text are in an average repo, but I think we can get closer to a metric for search functionality by saying "X KLOC per second" benchmarks (and this will also answer what the slow down is without talking about number of repos). If there is a scaling challenge unique to searching on 30K repos, then we should think about a metric that can track that.

Reply:
> I think the trouble with doing this by number of repos is that it conflates speed of structural/regex search and the performance of the infrastructure around dealing with 30K repos, without saying how big those repos are.

What if the OKR made a performance guarantee on structural search on k8s.sgdev.org (which is designed to replicate a large scale customer)? Regardless of the wording of this OKR, what we really what we want is "If current/future large customer X can use regular search, they should be able to use structural search" and by "use" I mean "searches complete in a reasonable amount of time, as judged by users". 

If we are having a hard time quantifying that, then lets change this KR to be about adding metrics so we can be more precise in the future.

Reply:
> What if the OKR made a performance guarantee on structural search on k8s.sgdev.org

I'd be onboard with that. The immediate question is the technical effort to just get a reliable signal for such a metric (benchmarking is hard, and if we do the whole end-to-end stack for a search query, there's bound to be jitter and other confounding variables). 

Thinking what this entails, it is more in the direction of 'change this KR to be about adding metrics', which would give us a clear signal to do the former. So :+1: for adding metrics.

Reply:
> I think the trouble with doing this by number of repos is that it conflates speed of structural/regex search and the performance of the infrastructure around dealing with 30K repos, without saying how big those repos are. 

This doesn't match my understanding exactly. Indexed search, for instance, allows search latency to **not** grow linearly with the number of repos mirrored by Sourcegraph.

Would there be ways to have that same sub-linear latency growth with structural search?

Challenges like repo size being uneven (i.e. work units aren't homogeneous) can be overcome — for instance by splitting repo archives into relatively equal sized parts and scaling infrastructure appropriately for parallel processing.

In the end we own the search problem end to end. As a user, what do you think should be the maximum amount of time you would wait to consider structural search "interactive", regardless of the number and individual size of repos?

> I'd be onboard with that. The immediate question is the technical effort to just get a reliable signal for such a metric (benchmarking is hard, and if we do the whole end-to-end stack for a search query, there's bound to be jitter and other confounding variables).

That's OK. What we want to optimize for is the user experience, so if there's jitter or other sources of slowness in the search path, we want to know about it and fix it. It's relatively easy to capture latency measurements over time with Prometheus, and we can always do before / after load tests, additionally.

Reply:
> Would there be ways to have that same sub-linear latency growth with structural search?

I understand the desire/vision here but can we even say that our unindexed regex search does not grow linearly by repos?

> As a user, what do you think should be the maximum amount of time you would wait to consider structural search "interactive", regardless of the number and individual size of repos?

I had a longish response to this, but I'm not going to answer it for sake of brevity. We can do that offline.

---

I think this can keep going back and forth, and in the interest of cutting that short, my overarching concern is "What does this KR mean?" If @nicksnyder' comment is accurate:

> I think the point of this KR is to assert that structural search can actually be used at large customers. 

and you think his suggestions are valid (I agree with them):

> What if the OKR made a performance guarantee on structural search on k8s.sgdev.org (which is designed to replicate a large scale customer)? 

> What if we changed this KR to be something like "add metrics for search latency per search type" (my assumption is that we don't track this from customer instances yet).

> we should probably add metrics regardless

then I feel that we need to establish and set a baseline and subsequent OKRs can iterate on that baseline. We can establish a baseline against, e.g., k8s. So this KR could be establishing a baseline performance for search along N parameters. I think this could apply to existing search generally, unless it's been done already or we believe it doesn't apply because "existing search is fast enough \<handwave\>"

---

Speaking very frankly, a much more impactful commitment than any of this would be introducing search pagination that gives us _more control_ to _display results quickly_ since it mimics a kind of streaming capability rather than optimizing our batch-like search path. If there is a mutual willingness from @slimsag and core to help accelerate that effort then we should allocate engineering to that instead.

Reply:
After catching up @tsenart and I agree that we'll revise this to emphasize "time to first result" for search. We'll stipulate that there should be a 99th percentile latency/time to return results within N seconds. This encompasses a wide possibility of strategies, including adding metrics, optimizing our end-to-end path (e.g., gitserver fetches and caches) and adding control knobs for displaying results faster with, e.g., pagination / streaming.